### PR TITLE
[gh-476] Sanitise HCL variables before storing on job submission

### PR DESCRIPTION
**Pull request by Juanadelacuesta**
### Description
Currently Nomad only handles HCL variables with new lines, any other non alphanumeric character is left untouched and stored unescaped, which can cause errors while re starting a stopped job, particularly from the UI.

it fixes [#476](https://github.com/hashicorp/terraform-provider-nomad/issues/476#issuecomment-2464341824)

### Contributor Checklist
- [x] **Changelog Entry** If this PR changes user-facing behavior, please generate and add a
  changelog entry using the `make cl` command.
- [x] **Testing** Please add tests to cover any new functionality or to demonstrate bug fixes and
  ensure regressions will be caught.
- [ ] **Documentation** If the change impacts user-facing functionality such as the CLI, API, UI,
  and job configuration, please update the  Nomad website documentation to reflect this. Refer to
  the [website README](../website/README.md) for docs guidelines. Please also consider whether the
  change requires notes within the [upgrade guide](../website/content/docs/upgrade/upgrade-specific.mdx).

### Reviewer Checklist
- [x] **Backport Labels** Please add the correct backport labels as described by the internal
  backporting document.
- [x] **Commit Type** Ensure the correct merge method is selected which should be "squash and merge"
  in the majority of situations. The main exceptions are long-lived feature branches or merges where
  history should be preserved.
- [ ] **Enterprise PRs** If this is an enterprise only PR, please add any required changelog entry
  within the public repository.

### CODE REVIEW

File: api/jobs.go [hunk 1]
```diff
@@ -1061,9 +1061,7 @@ func (js *JobSubmission) Canonicalize() {
 	// characters to preserve them. This way, when the job gets stopped and
 	// restarted in the UI, variable values will be parsed correctly.
 	for k, v := range js.VariableFlags {
-		if strings.Contains(v, "\n") {
-			js.VariableFlags[k] = strings.ReplaceAll(v, "\n", "\\n")
-		}
+		js.VariableFlags[k] = url.QueryEscape(v)
```
Comment (tgross) [inline api/jobs.go:1064]: This fixes the Terraform provider and the CLI, because they both use this `api` package, but does the UI already have an equivalent to this fix? (cc @philrenaud )

Should we add a warning to the API documentation for the [`JobSubmission` field](https://developer.hashicorp.com/nomad/api-docs/jobs#submission) that [third-party callers](https://developer.hashicorp.com/nomad/tools#community-tools) of the HTTP API need to perform this operation as well?

**Review by tgross — APPROVED** [review]
LGTM

**Review by jrasell — APPROVED** [review]
LGTM!

**Comment by philrenaud** [comment]
Testing this out made me remember we have two different variable-related properties in Job submission data: `VariableFlags`, which this PR addresses and are passed via CLI props, and `Variables`, which are what get entered via -var-file, etc.: https://github.com/hashicorp/nomad/blob/e206993d4990937225ef2c185809895ea7a357e9/api/jobs.go#L978-L984

I think that VariableFlags are affected here, but Variables are not. I'm not sure if the bug in question _could_ show up in that format, but wanted to mention it.

=========================

All that aside, I'm also noticing some perhaps erroneous character additions. Here are screenshots showing VariableFlags and output for a job where a cli-provided multi-line variable has a `+` character added:

on `main`:
Screenshot 2024-11-18 at 2 44 53 PM

on `f-NET-11421-vars`:
Screenshot 2024-11-18 at 2 42 37 PM

command to run was:

```
nomad job run -var meta_string="line1 \
\
\
\
line2" ../nomad-job-specs/unescaped_var_test.nomad.hcl
```

Note the `+` added between "line1" and "line2"

File: api/jobs.go (hunk 1 above)
Comment (philrenaud) [reply api/jobs.go:1064]: The UI runs output here through [this jsonToHcl() helper method](https://github.com/hashicorp/nomad/blob/main/ui/app/utils/json-to-hcl.js#L14-L24)

It adds `\n` between multiple VariableFlags / Variables entries, and the [codemirror editor](https://codemirror.net/) we use for rendering generally handles them correctly. It's not perfect, though: Your example in the test, 

```
"test": `"foo": "bar"`
```

renders like this:

image

which isn't perfect but it's probably enough of an edge case that it's been okay since we've had HCL-in-UI.

============

I'm not sure if this is what url.QueryEscape() is doing, but if we're given output like what's in the test below (%22foo%22%3A+%22bar%22), we will show that verbatim
image

**Review by philrenaud — APPROVED** [review]

**Review by tgross — APPROVED** [review]
LGTM!
